Use Accept header to specify response format

maxnikulin opened the pull request:
- Retain old `f=json` query parameter for backward compatibility
- Fix a typo in `server.request` iteration over headers

Add convenience arguments to `server.request()`:
- `csrfToken` to acquire token and add it to request,
- `accept` to specify response format,
- `form` to create `FormData` object and to add fields to it.

The intention is to allow authentication by an external application
that could obtain desired error format in a standard way #177.

danny0838 commented:
There are too many things intermingled together. Please separate them to different commits or patches.

1. Typo fix. (I don't see where it actually is yet)
2. Take js object for request body.
3. Take csrfToken to auto-fetch and append related form data.
4. Send Accept header rather than f=json.

We need further investigation about this request, however. Although the current implementation expects JSON, it can safely fallback if the response format doesn't comply, especially for error responses that the response body is only informative.

danny0838 commented:
Implemented some changes in the latest devel branch. You can rebase on this or wait for our further implementation.

maxnikulin commented:
> Implemented some changes in the latest devel branch. You can rebase on this or wait for our further implementation.

Sorry, did not get you point. Do you find it acceptable (at least some parts) and ask to resolve rebase conflicts?

> Typo fix. (I don't see where it actually is yet)

Purist's approach is to fix typo before other changes, but I made same mistake for the `form` argument and noticed it only after indentation of original line was changed and I was lazy to resolve that conflict during commit reordering. I do not see any usage of `headers` argument, but iteration over `Object.entries()` result should be done with `for-of`, not `for-in` loop.

As to other changes, they are scattered too widely, that is why it was a single commit. It is possible to split `f=`/`accept` changes post hoc. On the other `csrfToken` and `form` are coupled too tightly, and risk to introduce errors is rather high in comparison to little benefit.

danny0838 commented:
> Purist's approach is to fix typo before other changes, but I made same mistake for the `form` argument and noticed it only after indentation of original line was changed and I was lazy to resolve that conflict during commit reordering.

Although I usually do, I don't insist on commit reordering for contributors. Nevertheless, separating different topics into different commits/branches is another story. Mixing everything together makes review difficult, and forces us choosing from accept-all and reject-all instead of cherry-picking some of them and request a change for others.

Anyway, we already implemented some related things and there's already a conflict introduced. As there may be other related and possibly conflicting changes in the near future, and we already know how to implement it, this can be lazily left as-is. But if you still want to work on this, it's better separating the `accept` from others and rebase it on the latest devel branch, so that we can implement it easier if we want in the future.

maxnikulin commented:
Sorry, my bad, somehow I missed devel branch here and was surprised that there are no incompatible changes in PyWebScrapbook. I have extracted relevant changes to a separate commit.

danny0838 commented:
As a quick implementation, we can add an `expectJson` param for JSON requests, which both appends the `f=json` query and the accept header for the upper server to react, no matter whether PyWSB recognizes it or not.

The accept header for JSON should be something like `application/json, */*;q=0.1` to prevent a server response a 406 error.

danny0838 commented:
I think we can rename `accept` param to `format`.

We probably don't need the formCompatibility check, at least currently.

The accept header for JSON should be something like `application/json, */*;q=0.1` to prevent a server responding a 406 error.

For simplicity, just set `accept` header to which is defined by `format`

For the similar reason, maybe we should also append `f=json` by creating a URL and re-assign through SearchParams?

maxnikulin commented:
Renaming the parameter is not a problem, additional `*/*` is a bright idea.

> For simplicity, just set `accept` header to which is defined by `format`

Are you expecting `application/json, ...` value for each invocation (at least through constant defined somewhere) or I could keep the "mapping" inside `server.request()`?

I was not using `URL` object since there are direct operation with strings around.

I am thinking about moving of `f=` to the request body for POST queries with related addition to the property definition in backend.

danny0838 commented:
> > For simplicity, just set `accept` header to which is defined by `format`
> 
> Are you expecting `application/json, ...` value for each invocation (at least through constant defined somewhere) or I could keep the "mapping" inside `server.request()`?

Not sure about your implication. Currently it seems that only format=json need to have an accept header added.

> I was not using `URL` object since there are direct operation with strings around.

The difference is that setting via SearchParams overwrites existing `f` query params. This seems to be more consistent to the behavior of FormData and Header. It's not mandatory, though.

> I am thinking about moving of `f=` to the request body for POST queries with related addition to the property definition in backend.

We are not likely going to do that as it appears that keeping `f=...` in the query makes debugging through server logs easier.

maxnikulin commented:
I have renamed the argument field and switched to using of `URL` instead of direct string operations. `Accept` header value for JSON has been tuned as well.

danny0838 commented:
Did some tweak in the latest devel branch. Please check if it works and whether a further revision is needed.

danny0838 commented:
Tweaked and implemented in 0.78.0
